addons/microsoft_calendar/utils/microsoft_calendar.py

```diff
@@ -25,75 +23,159 @@ def wrapped(self, *args, **kwargs):
 class InvalidSyncToken(Exception):
     pass
 
+# In Outlook, an event can be:
+# - a 'singleInstance' event,
+# - a 'serie master' which contains all the information about an event reccurrence such as
+# - an 'occurrence' which is an event from a reccurrence (serie) that follows this reccurrence
+# - an 'exception' which is an event from a reccurrence (serie) but some differences with the reccurrence template (could be
+#   the name, the day of occurrence, ...)
+#
+#  All these kinds of events are identified by:
+#  - a event ID (id) which is specific to an Outlook calendar.
+#  - a global event ID (iCalUId) which is common to all Outlook calendars containing this event.
+#
+#  - 'singleInstance' and 'serie master' events are retrieved through the end-point `/v1.0/me/calendarView/delta` which provides
+#  the last modified/deleted items since the last sync (or all of these items at the first time).
+#  - 'occurrence' and 'exception' events are retrieved through the end-point `/v1.0/me/events/{serieMaster.id}/instances`,
+#  using the corresponding serie master ID.
+
 class MicrosoftCalendarService():
 
     def __init__(self, microsoft_service):
         self.microsoft_service = microsoft_service
 
     @requires_auth_token
-    def get_events(self, sync_token=None, token=None, timeout=TIMEOUT):
-        url = "/v1.0/me/calendarView/delta"
-        headers = {'Content-type': 'application/json', 'Authorization': 'Bearer %s' % token, 'Prefer': 'odata.maxpagesize=50,outlook.body-content-type="text"'}
-        params = {}
-        if sync_token:
-            params['$deltatoken'] = sync_token
-        else:
-            params['startDateTime'] = fields.Datetime.subtract(fields.Datetime.now(), years=3).strftime("%Y-%m-%dT00:00:00Z")
-            params['endDateTime'] = fields.Datetime.add(fields.Datetime.now(), years=3).strftime("%Y-%m-%dT00:00:00Z")
-        try:
-            status, data, time = self.microsoft_service._do_request(url, params, headers, method='GET', timeout=timeout)
-        except requests.HTTPError as e:
-            if e.response.status_code == 410 and 'fullSyncRequired' in str(e.response.content):
-                raise InvalidSyncToken("Invalid sync token. Full sync required")
-            raise e
-
+    def _get_events_from_paginated_url(self, url, token=None, params=None, timeout=TIMEOUT):
+        """
+        Get a list of events from a paginated URL.
+        Each page contains a link to the next page, so loop over all the pages to get all the events.
```

Comment:
Same remark. This method is public and disappear. We should discuss what we keep here and what we allow.
In the current state, the module is broken. But if somebody created a small module based on this one for simple cases, it could be an issue.

Reply:
`_get_events_from_paginated_url` is a new method called by the public method `get_events` which is still there ;-)